test/AutoDiff/downstream/tgmath_derivatives.swift.gyb

```diff
@@ -109,6 +109,15 @@ DerivativeTests.test("${op}_${T}") {
 %end
     }
   }
+
+  // pow
+  let dpow = ${op}(at: 3 as ${T}, 2 as ${T}, in: pow)
```

Comment:
Could you please add test cases where:
- `x` is negative, `y` is positive
- `x` is positive, `y` is negative
- `x` and `y` are both negative

---

`pow`'s gradient in Swift differs from Python when arguments are negative:
```python
import tensorflow as tf

def pow(x, y):
  x = tf.constant(x, dtype=tf.float32)
  y = tf.constant(y, dtype=tf.float32)
  with tf.GradientTape(persistent=True) as tape:
    tape.watch([x, y])
    z = tf.pow(x, y)
    dx, dy = tape.gradient(z, [x, y])
    print(dx.numpy(), dy.numpy())

for x in range(-3, 4):
  for y in range(-3, 4):
    pow(x, y)
pow(3, 2)
```
```swift
import Darwin

@derivative(of: pow)
func jvpPow(_ x: Float, _ y: Float) -> (value: Float, differential: (Float, Float) -> Float) {
  let value = pow(x, y)
  return (value, { dx, dy in
    dx * y * pow(x, y - 1) + dy * value * log(max(1, x))
  })
}

@derivative(of: pow)
func vjpPow(_ x: Float, _ y: Float) -> (value: Float, pullback: (Float) -> (Float, Float)) {
  let value = pow(x, y)
  return (value, { v in (
    v * y * pow(x, y - 1), v * value * log(max(1, x))
  )})
}

for i in -3...3 {
  for j in -3...3 {
    print(gradient(at: Float(i), Float(j), in: pow))
    // print(derivative(at: Float(i), Float(j), in: pow))
  }
}
```
<details>
<pre>
$ python3 pow.py
-0.037037037 -0.0
0.074074075 0.0
-0.11111111 -0.0
-0.0 0.0
1.0 -0.0
-6.0 0.0
27.0 -0.0
-0.1875 -0.0
0.25 0.0
-0.25 -0.0
-0.0 0.0
1.0 -0.0
-4.0 0.0
12.0 -0.0
-3.0 -0.0
2.0 0.0
-1.0 -0.0
-0.0 0.0
1.0 -0.0
-2.0 0.0
3.0 -0.0
-inf nan
-inf nan
-inf nan
nan 0.0
1.0 0.0
0.0 0.0
0.0 0.0
-3.0 0.0
-2.0 0.0
-1.0 0.0
0.0 0.0
1.0 0.0
2.0 0.0
3.0 0.0
-0.1875 0.0866434
-0.25 0.1732868
-0.25 0.3465736
0.0 0.6931472
1.0 1.3862944
4.0 2.7725887
12.0 5.5451775
-0.037037037 0.040689345
-0.074074075 0.12206803
-0.11111111 0.3662041
0.0 1.0986123
1.0 3.295837
6.0 9.88751
27.0 29.662533
6.0 9.88751
</pre>
</details>

<details>
<pre>
$ swift -Xfrontend -enable-experimental-cross-file-derivative-registration pow.swift
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(-inf, -nan)
(-inf, -nan)
(-inf, -nan)
(-nan, 0.0)
(1.0, 0.0)
(0.0, 0.0)
(0.0, 0.0)
(-3.0, 0.0)
(-2.0, 0.0)
(-1.0, 0.0)
(0.0, 0.0)
(1.0, 0.0)
(2.0, 0.0)
(3.0, 0.0)
(-0.1875, 0.0866434)
(-0.25, 0.1732868)
(-0.25, 0.3465736)
(0.0, 0.6931472)
(1.0, 1.3862944)
(4.0, 2.7725887)
(12.0, 5.5451775)
(-0.037037037, 0.040689345)
(-0.074074075, 0.12206803)
(-0.11111111, 0.3662041)
(0.0, 1.0986123)
(1.0, 3.295837)
(6.0, 9.88751)
(27.0, 29.662533)
</pre>
</details>

Reply:
Oh yes. Completely forgot to test negative values :/.

Reply:
I added test cases for combinations of positive and negative values, as well for the corner cases.

----

With the current implementation of `pow`'s derivative we can see that the gradients for negative values match the ones that your python script outputs above.

```swift
import Darwin

@derivative(of: pow)
func _vjpPow(_ x: Float, _ y: Float) -> (value: Float, pullback: (Float) -> (Float, Float)) {
  let value = pow(x, y)
  return (
    value,
    { v in
      (
        v * y * pow(x, y - 1), v * value * log(x.isLessThanOrEqualTo(0) ? Float(1) : x)
      )
    }
  )
}

for a in -3...3 {
    let x = Float(a)
    for b in -3...3 {
        let y = Float(b)
        print(gradient(at: x, y, in: pow))
    }
}
```
<details>
<summary>Show values</summary>

```
$> swift -Xfrontend -enable-experimental-cross-file-derivative-registration pow.swift
(-0.037037037, -0.0)
(0.074074075, 0.0)
(-0.11111111, -0.0)
(-0.0, 0.0)
(1.0, -0.0)
(-6.0, 0.0)
(27.0, -0.0)
(-0.1875, -0.0)
(0.25, 0.0)
(-0.25, -0.0)
(-0.0, 0.0)
(1.0, -0.0)
(-4.0, 0.0)
(12.0, -0.0)
(-3.0, -0.0)
(2.0, 0.0)
(-1.0, -0.0)
(-0.0, 0.0)
(1.0, -0.0)
(-2.0, 0.0)
(3.0, -0.0)
(-inf, -nan)
(-inf, -nan)
(-inf, -nan)
(-nan, 0.0)
(1.0, 0.0)
(0.0, 0.0)
(0.0, 0.0)
(-3.0, 0.0)
(-2.0, 0.0)
(-1.0, 0.0)
(0.0, 0.0)
(1.0, 0.0)
(2.0, 0.0)
(3.0, 0.0)
(-0.1875, 0.0866434)
(-0.25, 0.1732868)
(-0.25, 0.3465736)
(0.0, 0.6931472)
(1.0, 1.3862944)
(4.0, 2.7725887)
(12.0, 5.5451775)
(-0.037037037, 0.040689345)
(-0.074074075, 0.12206803)
(-0.11111111, 0.3662041)
(0.0, 1.0986123)
(1.0, 3.295837)
(6.0, 9.88751)
(27.0, 29.662533)
```
</details>